zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+    x2is: [F; 256],
+    nbits: [F; 256],
+}
+
+impl<W: PrimeField, N: PrimeField> Default for BlobKZGCircuit<W, N> {
+    fn default() -> Self {
+        BlobKZGCircuit {
+            input_values: InputValues::<W>::default(),
+            witness_values: WitnessValues::<W>::default(),
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<F: PrimeField> Default for InputValues<F> {
+    fn default() -> Self {
+        InputValues {
+            datapoints: [F::default(); NUMBER_OF_POINTS],
+            x: F::default(),
+            w: F::default(),
+            n: F::default(),
+        }
+    }
+}
+
+impl<F: PrimeField> Default for WitnessValues<F> {
+    fn default() -> Self {
+        WitnessValues {
+            evals: [F::default(); 4098],
+            wis: [F::default(); 4096],
+            xns: [F::default(); 257],
+            x2is: [F::default(); 256],
+            nbits: [F::default(); 256],
+        }
+    }
+}
+impl<W: PrimeField> WitnessValues<W> {
+    /// COMPUTATION OF PRIVATE WITNESS DATA
+    /// we have two main long computations:
+    /// a. the eval step
+    /// b. the xns step
+    ///
+    /// for the eval step we apply the main formula
+    /// to the datapoints, requiring the intermediate
+    /// rots of unity (wis) to be computed, as well as
+    /// the final exponentiation x^n (xns) to be computed.
+    ///
+    /// for the xns step we compute the square-and-multiply
+    /// algorithm, putting together all the possible powers
+    /// (x2is) to be computed and all the exponent's bits
+    /// (nbits) to be computed such that only the necessary powers
+    /// of x are multiplied together.
+    fn new(
+        InputValues {
+            datapoints,
+            x,
+            w,
+            n,
+        }: &InputValues<W>,
+    ) -> WitnessValues<W> {
+        let x2is: [W; 256] = {
+            let mut x2is = [W::ZERO; 256];
+            let mut state = *x;
+            for x in &mut x2is {
+                *x = state;
+                state *= state;
+            }
+            x2is
+        };
+        let nbits: [W; 256] = {
+            let mut nbits = [W::ZERO; 256];
+            let nbits_bool: [bool; 256] = {
+                let mut nbits_bool = [false; 256];
+                let repr: [u8; 32] = n.to_repr().as_ref().try_into().unwrap();
+                for (j, byte) in repr.iter().enumerate() {
+                    for i in 0..8 {
+                        nbits_bool[j * 8 + i] = ((byte >> i) & 1) == 1;
+                    }
+                }
+                nbits_bool
+            };
+
+            for (x, b) in nbits.iter_mut().zip(nbits_bool.iter()) {
+                if *b {
+                    *x = W::ONE
+                }
+            }
+            nbits
+        };
+        let xns: [W; 257] = {
+            let mut xns = [W::ZERO; 257];
+            let mut state = W::ONE;
+            for (x, (x2i, nbit)) in xns.iter_mut().zip(x2is.iter().zip(nbits.iter())) {
```

Comment:
nit: consider just using `i: 0..=256` instead of iterators to make it easier to read. here, we implicitly skip the last iteration because x2is will yield elements up to 256